nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp

```diff
@@ -372,6 +378,45 @@ void RegulatedPurePursuitController::rotateToHeading(
   angular_vel = std::clamp(angular_vel, min_feasible_angular_speed, max_feasible_angular_speed);
 }
 
+geometry_msgs::msg::Point circleSegmentIntersection(
+  const geometry_msgs::msg::PoseStamped & p1,
+  const geometry_msgs::msg::PoseStamped & p2,
+  double r)
+{
+  auto x1 = p1.pose.position.x;
+  auto x2 = p2.pose.position.x;
+  auto y1 = p1.pose.position.y;
+  auto y2 = p2.pose.position.y;
+
+  auto dx = x2 - x1;
+  auto dy = y2 - y1;
+  auto dr2 = std::pow(dx, 2) + std::pow(dy, 2);
+  auto D = x1 * y2 - x2 * y1;
+
+  auto intersection = [&](double sign) {
+      geometry_msgs::msg::Point p;
+      // Formula for intersection of a line with a circle centered at the origin:
+      // https://mathworld.wolfram.com/Circle-LineIntersection.html
+      // This works because the poses are transformed into the robot frame
+      double sqrt_term = std::sqrt(std::pow(r, 2) * dr2 - std::pow(D, 2));
+      p.x = (D * dy + sign * std::copysign(1.0, dy) * dx * sqrt_term) / dr2;
+      p.y = (-D * dx + sign * std::abs(dy) * sqrt_term) / dr2;
+      return p;
+    };
+  auto [x_min, x_max] = std::minmax(x1, x2);
+  auto [y_min, y_max] = std::minmax(y1, y2);
+  auto in_segment = [&](geometry_msgs::msg::Point p) {
+      return x_min <= p.x && p.x <= x_max &&
+             y_min <= p.y && p.y <= y_max;
+    };
+  auto intersection1 = intersection(1.0);
+  if (in_segment(intersection1)) {
```

Comment:
This can just be `if (x_min <= p.x && p.x <= x_max && y_min <= p.y && p.y <= y_max)` adding the lambda doesn't really make this more clear if its only used in 1 location

Reply:
Great point: I think that is an artifact of an earlier revision.

- [ ] Refactor `in_segment` lambda